astropy/units/quantity.py

```diff
@@ -685,7 +683,7 @@ def isscalar(self):
             (e.g. ``np.array(1)``), while this is True for quantities,
             since quantities cannot represent true numpy scalars.
         """
-        return not isiterable(self.value)
+        return not self.shape
```

Comment:
Your suggestion would make `isscalar` return exactly which it does now (but still faster), but I felt this was actually not great because it was inconsistent with the fact that for `numpy` a single item of a structured array is considered a scalar (i.e., `np.isscalar(np.array([(1., 1.)], 'f4,f4')[0]) is True`). So, I thought I should change it. This does imply the extra test in `__getitem__` but this will not be run often.

But this is a change from what `isiterable` gives, so this lead to issue that that `__getitem__` wouldn't work on the fields of an structured scalar Quantity.